card/lib/card/location.rb

```diff
@@ -0,0 +1,45 @@
+# -*- encoding : utf-8 -*-
+# helper for urls, links, redirects, and other location related things.
+#  note: i'm sure this isn't the optimal name..
+
+module Card::Location
+
+ # -----------( urls and redirects from application.rb) ----------------
+
+  # TESTME
+  def page_path title, opts={}
+    
+    format = opts[:format] ? ".#{opts.delete(:format)}"  : ''
+    query  = opts.present? ? "?#{opts.to_param}"         : ''
+    card_path "#{title.to_name.url_key}#{format}#{query}"
+  end
+
+  def card_path rel #should be in smartname?
+    rel_path = Card===rel ? rel.cardname.url_key : rel.to_s
+    if rel_path =~ /^\//
+      rel_path
+    else
+      "#{ Cardio.config.relative_url_root }/#{ rel_path }"
+    end
+  end
+
+  def card_url rel #should be in smartname?
+    if rel =~ /^https?\:/
+      rel
+    else
+      "#{ Card::Env[:protocol] }#{ Card::Env[:host] }#{ card_path rel }"
+    end
+  end
+
+
+  # Links ----------------------------------------------------------------------
+
+  def link_to_page( text, title=nil, options={})
+    title ||= text
+    url_options = {}
+    [:type, :view].each { |k| url_options[k] = options.delete(k) if options[k] }
+    url = page_path( title, url_options )
+    link_to text, url, options
+  end
+
```

Comment:
the history stuff is effectively a separate system; I doubt that separation will cause much problem.

the main overlap in the handling in controllers/formats is in all the wagn_url and wagn_path stuff.  the controller needs it for redirection/success stuff, and the formats need it for urls in content.

Yes, you could refer to the mods version from the controller, but from unless you're thinking of a bigger change than I know, it would have to be @card.format.method (not Card.method).

The nice thing about the mods approach is that you can then customize the handling of locations on a set basis. Can't say that's come up before, but it could prove useful at some point.